farmbot_core/lib/farmbot_core/asset/command.ex

```diff
@@ -2,21 +2,68 @@ defmodule FarmbotCore.Asset.Command do
   @moduledoc """
   A collection of functions that _write_ to the DB
   """
+  require Logger
 
-  alias FarmbotCore.{
-    Asset,
-    Asset.Repo,
-    Asset.Device
-  }
-
-  @type kind :: Device | FbosConfig | FwConfig
-  @type return_type ::
-          Device.t() | FbosConfig.t() | FwConfig.t()
-
-  @callback update(kind, params :: map()) :: return_type
-  def update(Device, params) do
-    Asset.device()
-    |> Device.changeset(params)
-    |> Repo.update!()
+  alias FarmbotCore.{Asset, Asset.Repo, Asset.Device, BotState}
+  alias FarmbotExt.API.{EagerLoader}
+  @type kind :: String.t()
+  @type params :: map()
+  @type id :: float()
+  @type return_type :: :ok
+
+  @callback update(kind, params, id) :: return_type
+  def update("Device", params) do
+    Asset.device() |> Device.changeset(params) |> Repo.update!()
+  end
+
+  def update("FbosConfig", params, _),
+    do: Asset.update_fbos_config!(params)
+
+  def update("FirmwareConfig", params, _),
+    do: Asset.update_firmware_config!(params)
+
+  def update("FarmwareEnv", params, id),
+    do: Asset.upsert_farmware_env_by_id(id, params)
+
+  def update("FarmwareInstallation", id, params),
+    do: Asset.upsert_farmware_env_by_id(id, params)
+
+  # Deletion use case:
+  def update(asset_kind, nil, id) do
+    old = Repo.get_by(as_module(asset_kind), id: id)
+    old && Repo.delete!(old)
+    :ok
+  end
+
+  # Catch-all use case:
+  def update(asset_kind, params, id) do
+    Logger.info("autosyncing: #{asset_kind} #{id} #{inspect(params)}")
+
+    case Repo.get_by(as_module(asset_kind), id: id) do
+      nil ->
+        struct(asset_kind)
+        |> asset_kind.changeset(params)
+        |> Repo.insert!()
+
+      asset ->
+        asset_kind.changeset(asset, params)
+        |> Repo.update!()
+    end
+
+    :ok
+  end
+
+  @callback cached_update(kind, params, id) :: return_type
+  def cached_update(asset_kind, params, id) do
+    mod = as_module(asset_kind)
+    asset = Repo.get_by(mod, id: id) || struct(asset_kind)
+    changeset = mod.changeset(asset, params)
+    :ok = EagerLoader.cache(changeset)
+    :ok = BotState.set_sync_status("sync_now")
+  end
+
+  # Convert string `"Device"` to module `Asset.Device`
+  defp as_module(asset_kind) do
+    Module.concat([Asset, asset_kind])
```

Comment:
You should probably throw an exception if this module doesn't exist, or have some other check. This is technically an atom leak since `Module.concat/2` with arbitrary strings will create an atom.

```elixir
defp as_module(kind) when kind in @known_str_kinds
...
```
or you can use a pattern match head:
(i like this one since it eliminates the possibility of runtime exceptions)
```elixir
defp as_module("Device"), do: FarmbotCore.Asset.Device
```

or you can do a slightly less safe:
```elixir
defp as_module(asset_kind) do
   module = Module.concat([Asset, asset_kind])
   true = Code.ensure_loaded(module)
   module
end
```